# Data Cleaning with Jupyter Notebook - 2009 to 2017 Dataset

In [1]:
# Importing necessary libraries

import pandas as pd
import numpy as np

# Loading the acquired dataset

df = pd.read_csv("NFL Play by Play 2009-2017 (v4).csv")

# Set seed for reproducibility

np.random.seed(0)

C:\Users\MY PC\AppData\Local\Temp\ipykernel_3064\740993034.py:8: DtypeWarning: Columns (25,51) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("NFL Play by Play 2009-2017 (v4).csv")


## Take a First Look at the Data

In [2]:
df.head()

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
0  2009-09-10  2009091000      1    1   NaN  15:00         15    3600.0   
1  2009-09-10  2009091000      1    1   1.0  14:53         15    3593.0   
2  2009-09-10  2009091000      1    1   2.0  14:16         15    3556.0   
3  2009-09-10  2009091000      1    1   3.0  13:35         14    3515.0   
4  2009-09-10  2009091000      1    1   4.0  13:27         14    3507.0   

   PlayTimeDiff SideofField  ...    yacEPA  Home_WP_pre  Away_WP_pre  \
0           0.0         TEN  ...       NaN     0.485675     0.514325   
1           7.0         PIT  ...  1.146076     0.546433     0.453567   
2          37.0         PIT  ...       NaN     0.551088     0.448912   
3          41.0         PIT  ... -5.031425     0.510793     0.489207   
4           8.0         PIT  ...       NaN     0.461217     0.538783   

   Home_WP_post  Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.546433      0.453567  0.485675  0.060758       NaN       NaN    2009  
1      0.551088      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.510793      0.489207  0.551088 -0.040295       NaN       NaN    2009  
3      0.461217      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.558929      0.441071  0.461217  0.097712       NaN       NaN    2009  

[5 rows x 102 columns]

In [3]:
df.tail()

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
407683  2017-12-31  2017123101     29    4   NaN  00:28          1      28.0   
407684  2017-12-31  2017123101     29    4   3.0  00:28          1      28.0   
407685  2017-12-31  2017123101     29    4   4.0  00:24          1      24.0   
407686  2017-12-31  2017123101     30    4   1.0  00:14          1      14.0   
407687  2017-12-31  2017123101     30    4   NaN  00:00          0       0.0   

        PlayTimeDiff SideofField  ...    yacEPA  Home_WP_pre  Away_WP_pre  \
407683           4.0         BAL  ...       NaN          NaN          NaN   
407684           0.0         BAL  ... -0.397515     0.080409     0.919591   
407685           4.0         BAL  ...  2.457114     0.050478     0.949522   
407686          10.0         BAL  ...       NaN     0.030881     0.969119   
407687          14.0         BAL  ...       NaN     0.000000     1.000000   

        Home_WP_post  Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  \
407683           NaN           NaN       NaN  0.000000       NaN       NaN   
407684      0.050478      0.949522  0.080409 -0.029931 -0.021795 -0.008136   
407685      0.030881      0.969119  0.050478 -0.019597 -0.030603  0.011006   
407686      0.000000      1.000000  0.969119  0.030881       NaN       NaN   
407687      0.000000      1.000000  0.999159  0.000000       NaN       NaN   

        Season  
407683    2017  
407684    2017  
407685    2017  
407686    2017  
407687    2017  

[5 rows x 102 columns]

The first you do when you upload a dataset into Jupyter Notebook is to see part of it. You can do so using the .head() or .tail() or any better way. 
This will help you to see if your data was loaded properly and identify any incosistencies. The NaN or None shows that that part of the dataset is a missing 
value.

# How many datapoints are missing?

In [4]:
# Checking the columns and rows we have in the data

df.shape

(407688, 102)

In [5]:
# Checking the number of missing data in each column
missing_data_counts = df.isnull().sum()

# We want to get the result from the first 10 columns
missing_data_counts[0:10]

Date                0
GameID              0
Drive               0
qtr                 0
down            61154
time              224
TimeUnder           0
TimeSecs          224
PlayTimeDiff      444
SideofField       528
dtype: int64

In [6]:
# Total missing values we have in the dataset
total_cells = np.product(df.shape)
total_missing = missing_data_counts.sum()

# Percentage missing data
percentage_missing_data = (total_missing/total_cells)*100
print(percentage_missing_data)

24.87214126835169


Therefore, approximately 25% of the data is missing. That is, we have about a quarter of the datapoints as missing values

## Figure out why the data is missing

The process of identifying why a datapoint is missing is referred to as data intution. This is the ability of the Data Scientist to guess the reason why a datapoint is not available. 
To perform data intution, it is important yourself: "Is the data missing because it was not recorded or it is empty because it is not possible to have an entry in that particular cell". For example, 
checking the weight of child of a woman who has no children is impossible. So in that case, it will be better to keep the cell as NaN. Otherwise, we will have to use what we call 'data imputation' to fill the missing value.

Let's work through an example. Looking at the number of missing values in the df dataframe, I notice that the column "TimesSec" has a lot of missing values in it:

In [9]:
missing_data_counts[0:30]

Date                  0
GameID                0
Drive                 0
qtr                   0
down              61154
time                224
TimeUnder             0
TimeSecs            224
PlayTimeDiff        444
SideofField         528
yrdln               840
yrdline100          840
ydstogo               0
ydsnet                0
GoalToGo            840
FirstDown         28811
posteam           24992
DefensiveTeam     24992
desc                  2
PlayAttempted         0
Yards.Gained          0
sp                    0
Touchdown             0
ExPointResult    397578
TwoPointConv     407083
DefTwoPoint      407664
Safety                0
Onsidekick            0
PuntResult       385317
PlayType              0
dtype: int64

It is important to read more about the dataset documentation of any given data. If you acquired the dataset from another person, do well to consult the person for more information about the dataset. This will give you an idea of what the data is about. You will also get to know whether a particular column that contains empty datapoints need to be filled or left like that because the values do not exist.

For instance, we see that the 224 missing datapoints of the TimeSecs was due to the fact that they were not recorded. Thus we must find values for these datapoints. On the other hand, the penalized column contains a lot datapoints that are empty due to the fact that they do not exist. That is, if you are not penalized, there is no way you will get a value recorded in that corresponding cell.

## Dropping missing values

This approach is not generally recommended for important projects! It's usually worth it to take the time to go through your data and really look at all the columns with missing values one-by-one to really get to know your dataset.

If after careful consideration, we decide to draw a particular row, we first use the pandas library and use the .dropna() function.

In [10]:
# dropping all rows that contains empty cells in our df dataset
df.dropna()

Empty DataFrame
Columns: [Date, GameID, Drive, qtr, down, time, TimeUnder, TimeSecs, PlayTimeDiff, SideofField, yrdln, yrdline100, ydstogo, ydsnet, GoalToGo, FirstDown, posteam, DefensiveTeam, desc, PlayAttempted, Yards.Gained, sp, Touchdown, ExPointResult, TwoPointConv, DefTwoPoint, Safety, Onsidekick, PuntResult, PlayType, Passer, Passer_ID, PassAttempt, PassOutcome, PassLength, AirYards, YardsAfterCatch, QBHit, PassLocation, InterceptionThrown, Interceptor, Rusher, Rusher_ID, RushAttempt, RunLocation, RunGap, Receiver, Receiver_ID, Reception, ReturnResult, Returner, BlockingPlayer, Tackler1, Tackler2, FieldGoalResult, FieldGoalDistance, Fumble, RecFumbTeam, RecFumbPlayer, Sack, Challenge.Replay, ChalReplayResult, Accepted.Penalty, PenalizedTeam, PenaltyType, PenalizedPlayer, Penalty.Yards, PosTeamScore, DefTeamScore, ScoreDiff, AbsScoreDiff, HomeTeam, AwayTeam, Timeout_Indicator, Timeout_Team, posteam_timeouts_pre, HomeTimeouts_Remaining_Pre, AwayTimeouts_Remaining_Pre, HomeTimeouts_Remaining_Post, AwayTimeouts_Remaining_Post, No_Score_Prob, Opp_Field_Goal_Prob, Opp_Safety_Prob, Opp_Touchdown_Prob, Field_Goal_Prob, Safety_Prob, Touchdown_Prob, ExPoint_Prob, TwoPoint_Prob, ExpPts, EPA, airEPA, yacEPA, Home_WP_pre, Away_WP_pre, Home_WP_post, Away_WP_post, Win_Prob, WPA, airWPA, ...]
Index: []

[0 rows x 102 columns]

This function removed all our dat. This is because there exists at least one empty cell in each row. Therefore we must try dropping columns that contain empty datapoints first.

In [11]:
# dropping columns that contains at least one empty cell
columns_with_na_dropped = df.dropna(axis=1)
columns_with_na_dropped.head()

Date      GameID  Drive  qtr  TimeUnder  ydstogo  ydsnet  \
0  2009-09-10  2009091000      1    1         15        0       0   
1  2009-09-10  2009091000      1    1         15       10       5   
2  2009-09-10  2009091000      1    1         15        5       2   
3  2009-09-10  2009091000      1    1         14        8       2   
4  2009-09-10  2009091000      1    1         14        8       2   

   PlayAttempted  Yards.Gained  sp  ...  Timeout_Indicator  Timeout_Team  \
0              1            39   0  ...                  0          None   
1              1             5   0  ...                  0          None   
2              1            -3   0  ...                  0          None   
3              1             0   0  ...                  0          None   
4              1             0   0  ...                  0          None   

   posteam_timeouts_pre HomeTimeouts_Remaining_Pre AwayTimeouts_Remaining_Pre  \
0                     3                          3                          3   
1                     3                          3                          3   
2                     3                          3                          3   
3                     3                          3                          3   
4                     3                          3                          3   

   HomeTimeouts_Remaining_Post  AwayTimeouts_Remaining_Post  ExPoint_Prob  \
0                            3                            3           0.0   
1                            3                            3           0.0   
2                            3                            3           0.0   
3                            3                            3           0.0   
4                            3                            3           0.0   

   TwoPoint_Prob  Season  
0            0.0    2009  
1            0.0    2009  
2            0.0    2009  
3            0.0    2009  
4            0.0    2009  

[5 rows x 41 columns]

In [14]:
# checking how much data we have lost
print("Columns in original dataset:", df.shape[1])
print("Columns after NaN's dropped:", columns_with_na_dropped.shape[1])
print("Total columns lost:", df.shape[1] - columns_with_na_dropped.shape[1])
print("Percentage columns lost:", ((df.shape[1] - columns_with_na_dropped.shape[1])/df.shape[1])*100)

Columns in original dataset: 102
Columns after NaN's dropped: 41
Total columns lost: 61
Percentage columns lost: 59.80392156862745


We have lost approximately 60% of our original columns. Howver, we have managed to get rid of all NaNs.

## Filling in missing values automatically

To see this clearly, let's first select a portion of the dataset for this operation. Here, we use the .loc[] tool.

In [15]:
subset_df = df.loc[:, 'EPA':'Season']
subset_df.head()

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474       NaN       NaN     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760       NaN       NaN     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796       NaN       NaN     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758       NaN       NaN    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295       NaN       NaN    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712       NaN       NaN    2009

We can use the Panda's fillna() function to fill in missing values in a dataframe for us. One option we have is to specify what we want the NaN values to be replaced with. Here, I'm saying that I would like to replace all the NaN values with 0.

In [17]:
#replace all NaNs with 0
subset_df.fillna(0).head()

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474  0.000000  0.000000     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  0.000000  0.000000     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796  0.000000  0.000000     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758  0.000000  0.000000    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.000000  0.000000    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712  0.000000  0.000000    2009

We can also fill the empty cells with whatever values comes right after them in the same column. (This makes a lot of sense for datasets where the observations have some sort of logical order to them.)

In [21]:
subset_df.fillna(method = 'bfill', axis = 0).head()

EPA    airEPA    yacEPA  Home_WP_pre  Away_WP_pre  Home_WP_post  \
0  2.014474 -1.068169  1.146076     0.485675     0.514325      0.546433   
1  0.077907 -1.068169  1.146076     0.546433     0.453567      0.551088   
2 -1.402760  3.318841 -5.031425     0.551088     0.448912      0.510793   
3 -1.712583  3.318841 -5.031425     0.510793     0.489207      0.461217   
4  2.097796 -0.343085  0.163935     0.461217     0.538783      0.558929   

   Away_WP_post  Win_Prob       WPA    airWPA    yacWPA  Season  
0      0.453567  0.485675  0.060758 -0.032244  0.036899    2009  
1      0.448912  0.546433  0.004655 -0.032244  0.036899    2009  
2      0.489207  0.551088 -0.040295  0.106663 -0.156239    2009  
3      0.538783  0.510793 -0.049576  0.106663 -0.156239    2009  
4      0.441071  0.461217  0.097712 -0.010456  0.006029    2009